protocols/abci/addressing-1174.md

```diff
@@ -0,0 +1,230 @@
+- _validity_: the decided block `v` satisfies the predefined predicate `valid(v)`.
+- _termination_: all correct processes eventually decide.
+
+The consensus algorithm implemented in CometBFT (Tendermint) guarantees these properties for each height.
+
+> To be a bit more formal, valid it a mathematical function here that depends only on its inputs.
+> The arxiv paper glosses over some details here. In fact, valid is a function of the proposed block `v`
+> and the state current state of the blockchain after applying the block of height `h-1`; let's write it as
+> bc(h-1). This we may write `valid(v, bc(h-1))`. Observe that this is a "mathematical function" that is
+> it only depends on its input. (In contrast to functions in progamming that may use other data to 
+> compute the result.
+
+### ABCI interface
+
+In the version of ABCI (`v0.17.0`) that existed before ABCI 1.0 and 2.0 (a.k.a. ABCI++),
+the implementation of the `valid(v)` predicate was totally internal to CometBFT.
+The application had no direct say on the validity of a block,
+although it could (and still can) influence it indirectly via the best-effort ABCI call `CheckTx`.
+
+With the evolution of ABCI to ABCI 1.0 and 2.0, the `valid(v)` predicate now has two components:
+
+- the validity checks performed directly by CometBFT on blocks
+  (block format, hashes, etc; the same as in ABCI `v0.17.0`)
+- the validity checks that now the application can perform as part of `ProcessProposal`
+
+With these new validity checks:
+
+- consensus _agreement_ is not affected and all processes are still required to agree on the same value..
+- the consensus _validity_ invariant is not affected since we are changing the
+  structure of the `valid(v)` predicate, on which consensus _validity_ directly depends.
+
+However, the new structure of `valid(v)` can affect _termination_ of consensus,
+as some implementations of `ProcessProposal` might reject values
+that CometBFT's internal validity checks would otherwise accept.
+
+This document focuses on how consensus _termination_ is affected
+by the new structure of `valid(v)`, in particular, the different implementations of `ProcessProposal`.
+
+### ABCI 1.0 (and 2.0) Specification
+
+The [ABCI 1.0 specification][abci-spec] imposes a set of new requirements to the application
+so that its implementation of `ProcessProposal` does not compromise _termination_ of consensus,
+given the current CometBFT consensus algorithm
+(called Tendermint, and described in the [arXiv paper][arxiv]).
+We reproduce a slightly modified version of these requirements here.
+This version of the requirements is equivalent to the ones provided in the ABCI 1.0 specification.
+The modifications are needed in order to express other requirements later on in this document.
+
+Let $p$ and $q$ be two correct processes.
+Let $r_p$ be a round of consensus at height $h$ where $p$ is the proposer.
+Let $s_{p,h-1}$ be $p$'s application's state committed for height $h-1$.
```

Comment:
A required assumption is that for every correct $p$ and $q$, $s_{p,h-1} = s_{q,h-1}$. This would allow us to remove the node indexer of the expression and render $s_{h-1}$ the same as the $bc(h-1)$ concept above proposed.

Reply:
I prefer to keep the same notation as in the ABCI spec. This way the reader (if they want) can compare them more easily